ambry-account/src/main/java/com/github/ambry/account/MySqlAccountService.java

```diff
@@ -108,28 +111,56 @@ Set<String> getRecentNotFoundContainersCache() {
   synchronized void fetchAndUpdateCache() throws SQLException {
     cachedAccountService.fetchAndUpdateCache();
     if (config.enableNewDatabaseForMigration) {
-      cachedAccountServiceNew.fetchAndUpdateCache();
+      try {
+        cachedAccountServiceNew.fetchAndUpdateCache();
+      } catch (SQLException e) {
+        if (config.ignoreNewDatabaseUploadError) {
+          logger.error("failed to do fetchAndUpdateCache for new db");
+        } else {
+          throw e;
+        }
+      }
     }
   }
 
   @Override
   public Account getAccountById(short accountId) {
+    if (config.enableGetFromNewDbOnly) {
+      return cachedAccountServiceNew.getAccountById(accountId);
+    }
     return cachedAccountService.getAccountById(accountId);
   }
 
   @Override
   public Account getAccountByName(String accountName) {
+    if (config.enableGetFromNewDbOnly) {
+      return cachedAccountServiceNew.getAccountByName(accountName);
+    }
     return cachedAccountService.getAccountByName(accountName);
   }
 
   @Override
   public boolean addAccountUpdateConsumer(Consumer<Collection<Account>> accountUpdateConsumer) {
-    return cachedAccountService.addAccountUpdateConsumer(accountUpdateConsumer);
+    boolean result = cachedAccountService.addAccountUpdateConsumer(accountUpdateConsumer);
+    if (config.enableNewDatabaseForMigration) {
+      boolean resultNew = cachedAccountServiceNew.addAccountUpdateConsumer(accountUpdateConsumer);
+      if (config.enableGetFromNewDbOnly) {
+        return resultNew;
+      }
+    }
+    return result;
   }
 
   @Override
   public boolean removeAccountUpdateConsumer(Consumer<Collection<Account>> accountUpdateConsumer) {
-    return cachedAccountService.removeAccountUpdateConsumer(accountUpdateConsumer);
+    boolean result = cachedAccountService.removeAccountUpdateConsumer(accountUpdateConsumer);
+    if (config.enableNewDatabaseForMigration) {
+      boolean resultNew = cachedAccountServiceNew.removeAccountUpdateConsumer(accountUpdateConsumer);
+      if (config.enableGetFromNewDbOnly) {
+        return resultNew;
+      }
+    }
+    return result;
```

Comment:
I probably missed these two methods in previous PR, dealing with account update consumers should be orthogonal to the migration. It should be a function of MysqlAccountService, instead of cached mysql store.

Reply:
it seems like the notifyAccountUpdateConsumers will be called in CacheAccountService::fetchAndUpdateCache.
So if we don't support addAccountUpdateConsumer, the accountUpdateConsumers in mysqlAccountService and CacheAccountService will be the different objects. So when we call notifyAccountUpdateConsumers in CacheAccountService::fetchAndUpdateCache, accountUpdateConsumers will be an empty list. Please let me know if I misunderstand anything.

Reply:
I think we should be careful about this one. Basically, this consumer callback is passed by some other components to get notified on the account updates. It's basically a passive read operation. If we can reading from old database, then we should only register this consumer callback to the old cached account service. If we are reading from the new database, then we should only register this consumer callback to the new cached account service.

It might be fine for the consumer to be called twice, once from old cached account service and the other one from the new, but I am afraid before data migration, these two invocation would read two different account lists.